### PR TITLE
MGMT-1577 auto assign host role 

### PR DESCRIPTION
**Pull request by filanov**
Still work in progress but it's enough to get the idea. 
We select the role only when user try to install and check if cluster is ready for installation before running any other validation.

**I still need to add tests, handle some edge cases and clean some duplication in the code **

### CODE REVIEW

File: internal/host/host.go [hunk 1]
```diff
@@ -22,6 +22,8 @@ import (
 	"github.com/thoas/go-funk"
 )
 
+const minMasterHosts = 3
```
Comment (avishayt) [inline internal/host/host.go:25]: Can we expose minHostsNeededForInstallation instead?

File: internal/host/host.go [hunk 2]
```diff
@@ -419,3 +414,66 @@ func (m *Manager) reportInstallationMetrics(ctx context.Context, h *models.Host,
...
+
+func (m *Manager) AutoAssignRole(ctx context.Context, h *models.Host, db *gorm.DB) error {
+	// select role if needed
+	if h.Role == models.HostRoleAutoAssign {
+		return m.autoRoleSelection(ctx, h)
+	}
+	return nil
+}
+
+func (m *Manager) selectRole(ctx context.Context, h *models.Host) (models.HostRole, error) {
```
Comment (avishayt) [inline internal/host/host.go:426]: Small nit: since the calls are AutoAssignRole -> autoRoleSelection -> selectRole, can you make the order they appear in the code the same?  Just makes reading the code a bit easier (i.e., move autoRoleSelection here right after AutoAssignRole).

File: internal/host/host.go [hunk 3]
```diff
@@ -419,3 +414,66 @@ func (m *Manager) reportInstallationMetrics(ctx context.Context, h *models.Host,
...
+		}
+	}
+
+	return autoSelectedRole, nil
+}
+
+func (m *Manager) autoRoleSelection(ctx context.Context, h *models.Host) error {
+	log := logutil.FromContext(ctx, m.log)
+	if h.Inventory == "" {
+		return errors.Errorf("host %s form cluster %s don't have hardware info",
```
Comment (avishayt) [inline internal/host/host.go:463]:
```suggestion
		return errors.Errorf("host %s from cluster %s don't have hardware info",
```

File: internal/host/host.go [hunk 4]
```diff
@@ -419,3 +414,66 @@ func (m *Manager) reportInstallationMetrics(ctx context.Context, h *models.Host,
...
+		if err != nil {
+			log.WithError(err).Errorf("failed to create new validation context for host %s", h.ID.String())
+			return autoSelectedRole, err
+		}
+		conditions, _, err := m.rp.preprocess(vc)
+		if err != nil {
+			log.WithError(err).Errorf("failed to run validations on host %s", h.ID.String())
+			return autoSelectedRole, err
+		}
+		if conditions[HasCPUCoresForRole] && conditions[HasMemoryForRole] {
```
Comment (avishayt) [inline internal/host/host.go:452]: It would be cleaner if we could run the code that transitions a host to known status with the role to see that it passes (in case we add a condition for known master later on and forget to add it here).

File: internal/host/host.go (hunk 1 above)
Comment (filanov) [reply internal/host/host.go:25]: i will move `minHostsNeededForInstallation` to common otherwise we will get cycle import issue

File: internal/cluster/ready.go [hunk 5]
```diff
@@ -34,14 +35,29 @@ func (r *readyState) RefreshStatus(ctx context.Context, c *common.Cluster, db *g
...
-	// Cluster is insufficient
-	mastersInKnown := mappedMastersByRole[intenralhost.HostStatusKnown]
-	if len(mastersInKnown) != minHostsNeededForInstallation {
-		log.Infof("Cluster %s dos not have exactly %d known master hosts, cluster is insufficient.", c.ID, minHostsNeededForInstallation)
-		return updateClusterStatus(log, db, *c.ID, swag.StringValue(c.Status), clusterStatusInsufficient, statusInfoInsufficient)
-
-		//cluster is still ready
-	} else {
-		return c, nil
+	knownMastersHostsCount := 0
```
Comment (ori-amizur) [inline internal/cluster/ready.go:38]: Isn't this duplicated code?

File: internal/host/host.go [hunk 6]
```diff
@@ -419,3 +414,66 @@ func (m *Manager) reportInstallationMetrics(ctx context.Context, h *models.Host,
...
+	}
+
+	if mastersCount < minMasterHosts {
+		h.Role = models.HostRoleMaster
+		vc, err := newValidationContext(h, m.db)
+		if err != nil {
+			log.WithError(err).Errorf("failed to create new validation context for host %s", h.ID.String())
+			return autoSelectedRole, err
+		}
+		conditions, _, err := m.rp.preprocess(vc)
```
Comment (ori-amizur) [inline internal/host/host.go:447]: Suggest to wrap this whole procedure "can-be-a-master" in a function

File: internal/host/host.go [hunk 7]
```diff
@@ -419,3 +414,66 @@ func (m *Manager) reportInstallationMetrics(ctx context.Context, h *models.Host,
...
+
+func (m *Manager) selectRole(ctx context.Context, h *models.Host) (models.HostRole, error) {
+	var (
+		autoSelectedRole = models.HostRoleWorker
+		log              = logutil.FromContext(ctx, m.log)
+	)
+
+	// count already existing masters
+	mastersCount := 0
+	if err := m.db.Model(&models.Host{}).Where("cluster_id = ? and status != ? and role = ?",
```
Comment (ori-amizur) [inline internal/host/host.go:434]: There are cases that a host does not have sufficient hardware to be even a worker.  How is that taken into account?

File: swagger.yaml [hunk 8]
```diff
@@ -2201,12 +2201,14 @@ definitions:
...
     type: string
     enum:
+      - 'auto-assign'
       - 'master'
       - 'worker'
 
   host-role:
     type: string
     enum:
+      - 'auto-assign'
```
Comment (ori-amizur) [inline swagger.yaml:2211]: I think that Swagger should consist from entities that are the building blocks that build the API.  auto-assign is an implementation detail of how it was decided to develop the this feature.  It is not a valid host-role and should not be part of the Swagger.
BTW, Isn't empty role is better? Why is it needed?

File: internal/bminventory/inventory.go [hunk 9]
```diff
@@ -669,17 +669,40 @@ func (b *bareMetalInventory) InstallCluster(ctx context.Context, params installe
...
 
-	if err = b.db.Preload("Hosts", "status <> ?", host.HostStatusDisabled).First(&cluster, "id = ?", params.ClusterID).Error; err != nil {
+	if err = b.db.Preload("Hosts", "status <> ?", host.HostStatusDisabled).
+		First(&cluster, "id = ?", params.ClusterID).Error; err != nil {
 		return common.NewApiError(http.StatusNotFound, err)
 	}
+	// auto select hosts roles if not selected yet.
+	err = b.db.Transaction(func(tx *gorm.DB) error {
+		// in case host monitor already updated the state we need to use FOR UPDATE option
+		transaction.AddForUpdateQueryOption(tx)
```
[the remark this answers is not in the document]
Comment (filanov) [reply internal/bminventory/inventory.go:679]: Change to what?

File: internal/cluster/insufficient.go [hunk 10]
```diff
@@ -44,14 +41,29 @@ func (i *insufficientState) RefreshStatus(ctx context.Context, c *common.Cluster
...
+		if swag.StringValue(h.Status) != models.HostStatusKnown &&
+			swag.StringValue(h.Status) != models.HostStatusDisabled {
+			return c, nil
+		}
+	}
+
 	// Cluster is ready
-	mastersInKnown, ok := mappedMastersByRole[models.HostStatusKnown]
-	if ok && len(mastersInKnown) == minHostsNeededForInstallation && c.APIVip != "" && c.IngressVip != "" {
+	if (knownMastersHostsCount == minHostsNeededForInstallation ||
```
[the remark this answers is not in the document]
Comment (filanov) [reply internal/cluster/insufficient.go:60]: It will be handled in a different PR. 
For now we just fail the install command.

File: internal/cluster/ready.go (hunk 5 above)
Comment (filanov) [reply internal/cluster/ready.go:38]: Not exactly. but as i wrote in the description it's still work in progress.

File: internal/host/host.go (hunk 7 above)
Comment (filanov) [reply internal/host/host.go:434]: It will be marked as insufficient worker.

File: swagger.yaml (hunk 8 above)
Comment (filanov) [reply swagger.yaml:2211]: User can change role to auto-assign so why is it not a building block?

File: internal/host/host.go (hunk 1 above)
Comment (filanov) [reply internal/host/host.go:25]: fixed, moved to `common/common.go` changed the name to `MinMasterHostsNeededForInstallation`

File: internal/host/host.go (hunk 4 above)
Comment (filanov) [reply internal/host/host.go:452]: Not sure i understand, role is not the only thing that set the host to known so not sure i understand your suggestion.

File: internal/bminventory/inventory.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (ori-amizur) [reply internal/bminventory/inventory.go:623]: Let the function return a pointer instead of assigning to an existing pointer

File: swagger.yaml (hunk 8 above)
Comment (ori-amizur) [reply swagger.yaml:2305]: auto-assign is not a real role.

Comment (avishayt) [reply swagger.yaml:2305]: @ori-amizur What is your suggestion?  I don't see a problem here.  The user can set the role to "auto-assign", and when they look at the host they will see "auto-assign".  Sure, that role isn't valid once the host starts the installation, but I don't see a problem with it.  And it is definitely more clear and explicit than empty role.

File: internal/bminventory/inventory.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (filanov) [reply internal/bminventory/inventory.go:623]: I don't understand the issue

File: internal/host/host.go (hunk 4 above)
Comment (avishayt) [reply internal/host/host.go]: Sorry I guess I didn't explain it clearly.  Maybe the comment is similar to Ori's about can-be-a-master.  My concern is that we might add another condition for master in the future and we will forget to update the code here.

File: internal/cluster/insufficient.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (filanov) [reply internal/cluster/insufficient.go]: Code was deleted

File: internal/host/host.go (hunk 6 above)
Comment (filanov) [reply internal/host/host.go]: Done

File: internal/cluster/ready.go (hunk 5 above)
Comment (filanov) [reply internal/cluster/ready.go]: File was deleted

File: internal/host/host.go (hunk 4 above)
Comment (filanov) [reply internal/host/host.go]: OK now i understand i will add another validator.

File: internal/bminventory/inventory.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (ori-amizur) [reply internal/bminventory/inventory.go:623]: https://github.com/openshift/assisted-service/blob/931bb06eedc3afadeb1f6c11881bb23b8c0a9189/pkg/transaction/transaction.go#L7
This is an unsafe implementation

[the remark this answers is not in the document]
Comment (filanov) [reply internal/bminventory/inventory.go:623]: Why? i just update the same object

[the remark this answers is not in the document]
Comment (filanov) [reply internal/bminventory/inventory.go:623]: if i do it like this it will be fine?
```
pdb := db.Set("gorm:query_option", "FOR UPDATE") 
*db = *pdb
```

[the remark this answers is not in the document]
Comment (ori-amizur) [reply internal/bminventory/inventory.go:623]: You are not allowed to copy contents of one gorm,DB object to another.  You can only copy pointers

File: swagger.yaml (hunk 8 above)
Comment (ori-amizur) [reply swagger.yaml:2305]: @avishayt I think empty role is a missing one.

File: internal/host/host.go (hunk 7 above)
Comment (ori-amizur) [reply internal/host/host.go]: OK

File: swagger.yaml (hunk 8 above)
Comment (filanov) [reply swagger.yaml:2305]: We don't have empty role anymore. 
The three options are auto-assign, worker, master
The default is auto-assign

**Comment by filanov** [comment]
test-infra passed http://10.46.10.2:8080/job/test-infra/2507/console

**Review by avishayt — APPROVED** [review]
